pandas/core/groupby/groupby.py

```diff
@@ -696,10 +696,18 @@ def _set_result_index_ordered(
         # the values/counts are repeated according to the group index
         # shortcut if we have an already ordered grouper
         if not self.grouper.is_monotonic:
-            index = Index(np.concatenate(self._get_indices(self.grouper.result_index)))
-            result.set_axis(index, axis=self.axis, inplace=True)
+            # row order is scrambled => sort the rows by position in original index
+            original_positions = Index(
+                np.concatenate(self._get_indices(self.grouper.result_index))
+            )
+            result.set_axis(original_positions, axis=self.axis, inplace=True)
             result = result.sort_index(axis=self.axis)
 
+            if len(original_positions) < len(self._selected_obj):
```

Comment:
I'm a bit confused in the case when this if statement does not hold (i.e. `==`). In this case, does the rest of this outer if block do anything given that `result.set_axis` is called below? This is also what happens in master, but I seem to be missing something.

Reply:
yes

on line 705 the index is the integer index that we used for sorting.

If no nans were dropped then the call to set_axis resets the original index. We want to go there if in that case because slicing an index can drop attributes (for example frequency) for some indexers. We have some tests that fail if we don't let that case fall through to set_axis

Reply:
Right - I understand that the call to `set_axis` on 711 is necessary in such a case. My question is: in the case where `set_axis` on 711 is called, does any other code within this function actually do anything? It seems to me the answer is "No", but perhaps I'm missing something.

Reply:
Ah - I found it. On line 704, the result is sorted. So there is a rearrangement there, this makes sense to me now.